### PR TITLE
Update bootstrap to v4.0.0

### PR DESCRIPTION
**Pull request by janslowik**
Becouse the bootstrap released the final stable 4 version i decided to update `uniforms-bootstrap4` from alpha-5 to version 4.0.0. I tested mostly on `demo` app and tried to keep similar appearance and behaviour like with alpha version. I tested also with some more complicated schemas  and imho looks good, but maybe something i missed so i will be happy on every review or remark.

### CODE REVIEW

File: packages/uniforms-bootstrap4/src/ErrorField.js [hunk 1]
```diff
@@ -6,8 +6,8 @@ import nothing        from 'uniforms/nothing';
 
 const Error = ({children, className, error, errorMessage, ...props}) =>
     !error ? nothing : (
-        <div className={classnames('card', className)} {...filterDOMProps(props)}>
-            <div className="card-block">
+        <div className={classnames('card', 'mb-3', className)} {...filterDOMProps(props)}>
```
Comment (radekmie) [inline packages/uniforms-bootstrap4/src/ErrorField.js]: Is this `mb-3` part really needed? We shouldn't enforce your layout.

It's also a few times more below.

File: packages/uniforms-bootstrap4/src/RadioField.js [hunk 2]
```diff
@@ -12,7 +12,8 @@ const Radio = props =>
                 className={classnames(
                     props.inputClassName,
                     'form-check',
-                    `radio${props.inline ? '-inline' : ''}` // bootstrap4 < alpha.6
+                    `radio${props.inline ? '-inline' : ''}`, // bootstrap4 < alpha.6
```
Comment (radekmie) [inline packages/uniforms-bootstrap4/src/RadioField.js]: IMHO we should get rid of these alpha fixes.

**Comment by radekmie** [comment]
@janslowik?

File: packages/uniforms-bootstrap4/src/ErrorField.js (hunk 1 above)
Comment (janslowik) [reply packages/uniforms-bootstrap4/src/ErrorField.js]: At version `4.0.0` `margin-bottom` was removed from some components. This `mb-3` class is added here to provide proper display the demo page. Without this submit buttons etc are stick to the form what imho looks bad.

File: packages/uniforms-bootstrap4/src/RadioField.js (hunk 2 above)
Comment (janslowik) [reply packages/uniforms-bootstrap4/src/RadioField.js]: done.

File: packages/uniforms-bootstrap4/src/BoolField.js [hunk 3]
```diff
@@ -10,7 +10,8 @@ const Bool = ({label, labelBefore, ...props}) =>
             className={classnames(
                 props.inputClassName,
                 'form-check',
-                `checkbox${props.inline ? '-inline' : ''}`, // bootstrap4 < alpha.6
+                `checkbox${props.inline ? ' custom-control-inline' : ''}`,
+                {'text-danger': props.error}
```
Comment (radekmie) [inline packages/uniforms-bootstrap4/src/BoolField.js]: In such case, use `'custom-control-inline': props.inline`.

File: packages/uniforms-bootstrap4/src/gridClassName.js [hunk 4]
```diff
@@ -3,32 +3,59 @@ import filterDOMProps from 'uniforms/filterDOMProps';
...
+        }
+    }
+
+    if (a === 'lg') {
+        if (b === 'xl') {
+            return -1;
+        }
+    }
+
+    return 1;
```
Comment (radekmie) [inline packages/uniforms-bootstrap4/src/gridClassName.js]: Do these have to be in order? If so, maybe let's have an array with correct order and order them by an index in this array: `sizes.indexOf(a) - sizes.indexOf(b)`?

File: packages/uniforms-bootstrap4/src/gridClassName.js [hunk 5]
```diff
@@ -38,7 +65,14 @@ export default function gridClassName (grid, side) {
 
     // Example: {xs: 6, sm: 4, md: 3}
     if (typeof grid === 'object') {
-        return Object.keys(grid).map(size => gridClassNamePart(size, grid[size], side)).join(' ');
+
```
Comment (radekmie) [inline packages/uniforms-bootstrap4/src/gridClassName.js]: Extra newline.

File: packages/uniforms-bootstrap4/src/gridClassName.js [hunk 6]
```diff
@@ -38,7 +65,14 @@ export default function gridClassName (grid, side) {
...
+        if (!grid['xs']) {
+            grid['xs'] = grid['sm'] || grid['md'] || grid['lg'] || grid['xl'];
+        }
```
Comment (radekmie) [inline packages/uniforms-bootstrap4/src/gridClassName.js]: Two things here:
* Should we care about it?
* If so, use dot notation instead.

File: packages/uniforms-bootstrap4/src/gridClassName.js (hunk 5 above)
Comment (janslowik) [reply packages/uniforms-bootstrap4/src/gridClassName.js]: It was intentionally.

File: packages/uniforms-bootstrap4/src/gridClassName.js (hunk 4 above)
Comment (janslowik) [reply packages/uniforms-bootstrap4/src/gridClassName.js]: ok.

Maybe?
```
const gridOrder = {
    xs: 1,
    sm: 2,
    md: 3,
    lg: 4,
    xl: 5
};

function compareSizeClass (a, b) {
    return gridOrder[a] - gridOrder[b];
}
```

File: packages/uniforms-bootstrap4/src/gridClassName.js (hunk 6 above)
Comment (janslowik) [reply packages/uniforms-bootstrap4/src/gridClassName.js]: Yes it is needed (becouse we use this `xs` to count the default `col-`.

Comment (radekmie) [reply packages/uniforms-bootstrap4/src/gridClassName.js]: It doesn't mean it's needed - it means that it's required.

File: packages/uniforms-bootstrap4/src/gridClassName.js [hunk 7]
```diff
@@ -38,7 +39,13 @@ export default function gridClassName (grid, side) {
 
     // Example: {xs: 6, sm: 4, md: 3}
     if (typeof grid === 'object') {
-        return Object.keys(grid).map(size => gridClassNamePart(size, grid[size], side)).join(' ');
+        if (!grid.xs) {
+            grid.xs = grid.sm || grid.md || grid.lg || grid.xl;
```
Comment (radekmie) [inline packages/uniforms-bootstrap4/src/gridClassName.js]: It just hit me: you can't mutate `grid` here!

File: packages/uniforms-bootstrap4/src/gridClassName.js [hunk 8]
```diff
@@ -38,7 +39,10 @@ export default function gridClassName (grid, side) {
 
     // Example: {xs: 6, sm: 4, md: 3}
     if (typeof grid === 'object') {
-        return Object.keys(grid).map(size => gridClassNamePart(size, grid[size], side)).join(' ');
+        const completeGrid = {xs: grid.sm || grid.md || grid.lg || grid.xl, ...grid};
```
Comment (radekmie) [inline packages/uniforms-bootstrap4/src/gridClassName.js]: Move it into an `if`. This way we'd make a lot of unnecessary copies.

Comment (janslowik) [reply packages/uniforms-bootstrap4/src/gridClassName.js]: Do you prefer
```
        if (!grid.xs) {
            const completeGrid = {xs: grid.sm || grid.md || grid.lg || grid.xl, ...grid};
            return Object.keys(completeGrid || grid)
                .sort(compareSizeClass)
                .map(size => gridClassNamePart(size, completeGrid[size], side)).join(' ');
        }

        return Object.keys(grid)
            .sort(compareSizeClass)
            .map(size => gridClassNamePart(size, grid[size], side)).join(' ');
```
or
```
const completeGrid = grid.xs
            ? grid
            : {xs: grid.sm || grid.md || grid.lg || grid.xl, ...grid};

        return Object.keys(completeGrid)
            .sort(compareSizeClass)
            .map(size => gridClassNamePart(size, completeGrid[size], side)).join(' ');
```

Comment (radekmie) [reply packages/uniforms-bootstrap4/src/gridClassName.js]: I think our ESLint allows overriding the argument. If so, then

```js
if (!grid.xs) {
    grid = {xs: grid.sm || grid.md || grid.lg || grid.xl, ...grid};
}
```

If not, then the second one.

**Pull request merged**